pgmax/groups/variables.py

```diff
@@ -1,73 +1,136 @@
+    num_states: Union[int, np.ndarray]
+
+    def __post_init__(self):
+        if np.prod(self.shape) > MAX_SIZE:
+            raise ValueError(
+                f"Currently only support NDVariableArray of size smaller than {int(MAX_SIZE)}. Got {np.prod(self.shape)}"
+            )
+
+        if np.isscalar(self.num_states):
+            num_states = np.full(self.shape, fill_value=self.num_states, dtype=np.int64)
+            object.__setattr__(self, "num_states", num_states)
+        elif isinstance(self.num_states, np.ndarray) and np.issubdtype(
+            self.num_states.dtype, int
+        ):
+            if self.num_states.shape != self.shape:
+                raise ValueError(
+                    f"Expected num_states shape {self.shape}. Got {self.num_states.shape}."
+                )
+        else:
+            raise ValueError(
+                "num_states should be an integer or a NumPy array of dtype int"
+            )
 
-    def _get_names_to_variables(
-        self,
-    ) -> OrderedDict[Union[int, Tuple[int, ...]], nodes.Variable]:
-        """Function that generates a dictionary mapping names to variables.
+        # Only compute the hash once, which is guaranteed to be an int64
+        this_id = id(self) % 2**32
+        this_hash = this_id * int(MAX_SIZE)
+        assert this_hash < 2**63
+        object.__setattr__(self, "this_hash", this_hash)
+
+    def __getitem__(
+        self, val: Union[int, slice, Tuple]
+    ) -> Union[Tuple[int, int], List[Tuple[int, int]]]:
+        """Given an index or a slice, retrieve the associated variable(s).
+        Each variable is returned via a tuple of the form (variable hash, number of states)
+
+        Note: Relies on numpy indexation to throw IndexError if val is out-of-bounds
+
+        Args:
+            val: a variable index or slice
 
         Returns:
-            a dictionary mapping all possible names to different variables.
+            A single variable or a list of variables
         """
-        names_to_variables: OrderedDict[
-            Union[int, Tuple[int, ...]], nodes.Variable
-        ] = collections.OrderedDict()
-        for name in itertools.product(*[list(range(k)) for k in self.shape]):
-            if len(name) == 1:
-                names_to_variables[name[0]] = nodes.Variable(self.num_states)
-            else:
-                names_to_variables[name] = nodes.Variable(self.num_states)
+        assert isinstance(self.num_states, np.ndarray)
+
+        if isinstance(val, slice) or (
+            isinstance(val, tuple) and isinstance(val[0], slice)
+        ):
+            assert isinstance(self.num_states, np.ndarray)
+            vars_names = self.variable_names[val].flatten()
+            vars_num_states = self.num_states[val].flatten()
+            return list(zip(vars_names, vars_num_states))
+
+        return (self.variable_names[val], self.num_states[val])
+
+    @cached_property
+    def variables(self) -> List[Tuple]:
```

Comment:
`Tuple[int, int]`

Reply:
that's not true in the case of VariableDict

Reply:
Out of curiosity, why don't we use the nicer `tuple[int, int]` form?

Reply:
We can make a global switch! We started with `Tuple` so for consistency has been using `Tuple`